### PR TITLE
mypy: tackle all violations in excluded files

### PR DESCRIPTION
**Pull request by booxter**
Next steps will be: adding more type annotations to existing functions; enabling stricter validations (--strict), revisiting existing type annotations and maybe tightening where possible (str -> Path, fewer Unions etc.)

### CODE REVIEW

**Comment by booxter** [comment]
The remaining violations in the file are tackled in https://github.com/instructlab/instructlab/pull/1287 Both PRs combined should allow us to enable mypy for this module.

**Review by leseb — DISMISSED** [review]

**Comment by booxter** [comment]
I'm collecting all relevant fixes for mypy here now. Seems counterproductive to split these in tiny PRs.

**Comment by booxter** [comment]
Converted to draft so that reviewers are aware this is ongoing work.

**Comment by booxter** [comment]
This is ready for review.

**Review by leseb — DISMISSED** [review]
Very nice, meticulous work! Congrats on clearing that list of violations!

File: src/instructlab/data/generator/generate_data.py
Comment (bjhargrave) [inline src/instructlab/data/generator/generate_data.py:7]: I am not sure I would bother changing files in this package since it is moving to the sdg repo. For now, I would leave the package in the exclude list.

**Review by bjhargrave — COMMENTED** [review]
There seems a lot of new code and deleted code in this PR which seems beyond linting fixes. I am unsure if those changes are safe.

**Comment by booxter** [comment]
> There seems a lot of new code and deleted code in this PR which seems beyond linting fixes. I am unsure if those changes are safe.

which specific cases do you mean? there was some dead code removed (that violates mypy). Beyond that, I don't think I touched the code.

File: src/instructlab/data/generator/generate_data.py
Comment (booxter) [reply src/instructlab/data/generator/generate_data.py:7]: We can do that too. I don't know the timeline of the migration of the code. We can move these changes to sdg (as part of mypy enablement there.) If it's the only concern, I can update the PR. But I'd like to see if any other specific issues are of concern.

File: src/instructlab/llamacpp/llamacpp_convert_to_gguf.py
Comment (bjhargrave) [inline src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:1616]: > which specific cases do you mean?

A param name change can break callers.

File: src/instructlab/llamacpp/llamacpp_convert_to_gguf.py
Comment (bjhargrave) [inline src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:1154]: > which specific cases do you mean?

This code is deleted. And above params.n_experts was deleted.

**Comment by booxter** [comment]
btw I split the PR into small commits, so we can discuss each of them individually. If you think a particular change is invalid, please point to the commit with a link so we discuss the same thing.

File: src/instructlab/mlx_explore/gguf_convert_to_mlx.py
Comment (bjhargrave) [inline src/instructlab/mlx_explore/gguf_convert_to_mlx.py:78]: > which specific cases do you mean?

This is now just 1 since rope_scale was removed from ModelArgs.

File: src/instructlab/mlx_explore/gguf_convert_to_mlx.py
Comment (bjhargrave) [inline src/instructlab/mlx_explore/gguf_convert_to_mlx.py:55]: > which specific cases do you mean?
 Not sure why this `or` was added or if the proper value is on the right hand side?

**Review by bjhargrave — COMMENTED** [review]
A couple of the code changes I was referring to.

File: src/instructlab/llamacpp/llamacpp_convert_to_gguf.py
Comment (booxter) [reply src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:1154]: Let me double check. I *think* that this code was dead because of how params are loaded, but I will confirm.

File: src/instructlab/llamacpp/llamacpp_convert_to_gguf.py
Comment (booxter) [reply src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:1616]: That's why the patch changes the callers where needed to adjust.

The reason for the change in variable name is because the function later reuses the same name to store a different type, which is a warning in mypy. (You are not advised to mix types like this.)

Comment (booxter) [reply src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:1616]: An alternative here can be a rename of the local variable instead. I don't have huge preference.

File: src/instructlab/llamacpp/llamacpp_convert_to_gguf.py
Comment (booxter) [reply src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:1154]: OK, the explanation as to why these two options are removed is here: 76410ae05cd81aeee305a48ed83da373d0ec28bf To quote,

```
    mypy: Remove use of undefined GGUFWriter symbols

    Neither add_expert_count nor add_expert_used_count are defined in the
    version of gguf library we use.  Actually, these symbols are not yet
    released on pypi, and are available in git repo of the library only:

    https://github.com/ggerganov/llama.cpp/commit/799a1cb13b0b1b560ab0ceff485caed68faa8f1f

    This makes mypy complain about it.

    This patch removes code related to these parameters. If and when we
    upgrade to gguf version that supports them, we may revert this patch.
```

Since GGUFWriter doesn't have the functions, passing these settings through config file would crash the code anyway. In the future, gguf may be updated to include these functions, then this may be revisited.

Let me know if this line of thought is ok for you.

File: src/instructlab/data/generator/generate_data.py
Comment (booxter) [reply src/instructlab/data/generator/generate_data.py:7]: It's unclear if you insist, or it's an opinion. For now, I will leave these changes included. But please let me know. (If so, could you please list the exact modules that are going to be migrated - or is the list captured somewhere?)

**Comment by booxter** [comment]
The latest is just a merge from `main`. No significant changes to the patch series itself.

File: src/instructlab/llamacpp/llamacpp_convert_to_gguf.py
Comment (bjhargrave) [reply src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:1154]: I guess I am worried were are changing 3rd party files just to make mypy happy. Perhaps we should exclude these 3rd party source files (e.g. this and the mlx_explore folder) from mypy checks.

Comment (booxter) [reply src/instructlab/llamacpp/llamacpp_convert_to_gguf.py:1154]: Depends on how you look at it. It's not just to make `mypy` happy: this patch removes a bug that results in a crash depending on what's in the config file. With this patch included, the unknown options will be silently ignored.

I was wondering about the nature of this file. The way these files are managed in this tree is ambiguous. On one hand, they clearly come from `llamacpp` repo. On the other hand, it's not pulled in through a library (which makes sense - since this code is NOT part of their package); and we don't track which particular version of the file we pulled here. (I actually tried to figure this out - and I suspect what we have here in-tree is a **modified** version of one of older versions of the `llamacpp` example file.)

We already fixed some bugs in this file, removed `click` code, and `blackened` / `ruffed` the file. (Just as examples. There are more.) If the decision is to not touch the file (and maybe some other files - which are these?), then we should document this (there's a README uder this path, we could update it.) Ideally, we would also document where this code comes from then (the exact commit message), or maybe even try to pull it through git submodules. (Not sure it's possible at this point though - the code diverged.)

(My personal opinion is that this would not be a good decision to make, because this code already diverged from wherever we took it from. But if the team feels otherwise, we could try to freeze new changes to the module, somehow, at least as a review policy.)

File: src/instructlab/data/generator/generate_data.py
Comment (booxter) [reply src/instructlab/data/generator/generate_data.py]: These are gone now in the latest tree.

File: src/instructlab/llamacpp/llamacpp_convert_to_gguf.py
Comment (booxter) [reply src/instructlab/llamacpp/llamacpp_convert_to_gguf.py]: @bjhargrave all patches for the modules you identified (plus one more) that are copy-pasted are dropped, series flattened. This should be less controversial, I think.

File: src/instructlab/utils.py
Comment (bjhargrave) [inline src/instructlab/utils.py:366]:
```suggestion
        raise NoSuchResource(str(path)) from e
```

This will make mypy happy.

File: src/instructlab/utils.py
Comment (bjhargrave) [inline src/instructlab/utils.py:341]: This really is a Traversable. It just moved in Python 3.11.

```python
try:
    from importlib.resources.abc import Traversable  # type: ignore[import-not-found]
except ImportError:  # python<3.11
    from importlib.abc import Traversable
```

We should use the proper type rather than coerce it to Path.

File: requirements.txt
Comment (bjhargrave) [inline requirements.txt:49]: This should be brought in by `jsonschema`. Not sure why it should be explicit here.

File: src/instructlab/utils.py
Comment (bjhargrave) [inline src/instructlab/utils.py:397]: Why do this? It is not necessary except to satisfy the change to use Path from Traversable above.

File: src/instructlab/utils.py
Comment (bjhargrave) [inline src/instructlab/utils.py:548]: Not sure why this is necessary? The schema validation has already made sure the key is present.

**Review by bjhargrave — CHANGES_REQUESTED** [review]
Much simpler with the removal of the copied files!

File: requirements.txt
Comment (booxter) [reply requirements.txt:49]: I assume that if it's directly imported, then it's a direct dependency.

File: src/instructlab/utils.py
Comment (booxter) [reply src/instructlab/utils.py:341]: You can see all reasons for each change documented in commit messages, e.g. here for this change: https://github.com/instructlab/instructlab/pull/1275/commits/344f697b739f12771b9e1a6de4b846b775a04805

The reason why this has to change to Path is because we apply LRU cache decorator that expects hashable arguments, and Traversable is not hashable.

File: src/instructlab/utils.py
Comment (booxter) [reply src/instructlab/utils.py:397]: Yes, it's to satisfy the other change, which is needed to fix type violation where `@cache` decorator is applied to a function that receives (potentially) non-hashable arguments (`Traversable`). Both changes belong to the same commit.

File: src/instructlab/utils.py
Comment (booxter) [reply src/instructlab/utils.py:548]: Yes, though this fact is not known to type checking layer. `mypy` reasons that the field **may** be missing (e.g. a bug in schema validation module may expose it.)

**Review by leseb — APPROVED** [review]
Thanks a lot for taking care of this.

**Comment by tiran** [comment]
I forgot about this PR...

A few suggestions:
- We dropped support for older versions of Python and require at least 3.10. You can now use `str | None` instead of `Optional[str]`. I'm appending a Ruff configuration change that lets you auto-fix the code.
- IMHO the PR is fixing too much at once. I really, really want proper type information in InstructLab. But a big patch is harder to land and disruptive for others. Do you have an idea how to split the PR into smaller chunks? Maybe address one module at a time?
- InstructLab mixes and matches str and pathlib.Path for paths. We should use `pathlib.Path` consistently for all paths and start with CLI args. Clicks `click.Path` type does not return a `pathlib.Path`. I created a custom `ClickPath` object for Fromager. Please take a look at https://github.com/python-wheel-build/fromager/pull/112
- Personal opinion: I don't like `from pathlib import Path` or `from typing import Any`. I prefer `import pathlib` and `pathlib.Path` for readability. Can I persuade you to avoid from imports? :) 

<details>
<summary>Auto-fixer for type annotations</summary>

```
diff --git a/pyproject.toml b/pyproject.toml
index 3484a10..fe441bf 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -68,7 +68,7 @@ include = [
 ignore = ["W002"]
 
 [tool.ruff]
-target-version = "py39"
+target-version = "py310"
 # same as black's default line length
 line-length = 88
 
@@ -76,16 +76,21 @@ line-length = 88
 # Allow fix for all enabled rules (when `--fix`) is provided.
 fixable = ["ALL"]
 unfixable = []
+# modern type annotations
+extend-safe-fixes = ["UP006", "UP007"]
 
 # Fixers will be enabled gradually.
 select = [
   # "B",  # flake8-bugbear
   # "E",  # pycodestyle
   # "F",  # Pyflakes
+  "F401", # unused imports
   "Q", # flake8-quotes
   # Ruff does not support isort's import_headings feature, yet.
   # "I",  # isort
   # "UP",  # pyupgrade
+  "UP006",
+  "UP007",
   # "SIM",  # flake8-simplify
   "TID", # flake8-tidy-imports
 ]
```

</details>

**Comment by booxter** [comment]
@tiran I appreciate the feedback.

The series started as separate PR per module patched, but these never went anywhere, so I figured that at least once I get some attention from reviewers, I will have a chance to land it before yet another git conflict. Technically, it's simple to split these, since I already split each change into a separate commit. But I don't see how this will get us anywhere, except into months of struggling for reviewer attention.

Re: Optional vs None, from xxx import yyy, and path types - we can follow up. I want to follow up with path type switches, but I am holding it off for when we enable the checks for the whole tree and stop adding more excludes to the list instead of fixing legit failures. (Which seems to be a habit here.) I don't see how squashing more changes like this here will get us quicker to where we could at least start enforcing "no new mypy violations" for new patches.

File: requirements.txt
Comment (tiran) [reply requirements.txt:48]: Don't pin upper version unless a higher version introduces a known breaking change.

File: src/instructlab/utils.py
Comment (tiran) [inline src/instructlab/utils.py:309]: I think this should be a normal `str`.

File: src/instructlab/utils.py
Comment (tiran) [inline src/instructlab/utils.py:381]: Not your fault, but oh boy, these three lines are unreadable.

File: src/instructlab/utils.py
Comment (tiran) [inline src/instructlab/utils.py:387]:
```suggestion
        with file_path_p.open(encoding="utf-8") as file:
```

File: src/instructlab/utils.py
Comment (tiran) [reply src/instructlab/utils.py:311]: This code can break the cache under some conditions. `resource.as_file` can point to a random temporary file instead of a stable file name.
I recommend to leave the code as it for now. Schema loading and validation should be moved to the `instructlab.schema` package. The package should use an explicit cache with key cache `(version, basename)`.

**Review by leseb — APPROVED** [review]
@booxter thanks for your patience while working on this. I'd like to move forward with this, it has been around for quite some time and the constant rebase/conflict makes it painful I'm sure. I think @tiran has small nits but we should be able to sort it out rapidly. Thanks again.

**Comment by tiran** [comment]
Another nit pick is: This PR has 35 commits. InstructLab project does not use squash commits. I don't think it's a good idea to add 35 commits + 1 merge commit to the git history of main branch. It's way too noisy.

File: src/instructlab/utils.py
Comment (booxter) [reply src/instructlab/utils.py:272]: I am happy to do it. But this would have nothing to do with mypy enablement. I will make a note to follow up with cleaning up these pylint exceptions separately.

File: src/instructlab/utils.py
Comment (bjhargrave) [reply src/instructlab/utils.py:257]: I have fixed this in https://github.com/instructlab/schema/pull/33.

That PR is to centralize the taxonomy parsing code into the instruct lab-schema package which can be used by other package which need to parse qna.yaml files.

File: src/instructlab/utils.py
Comment (booxter) [reply src/instructlab/utils.py:311]: I'll replace with line-specific ignore comment for now.

**Comment by booxter** [comment]
> I don't think it's a good idea to add 35 commits + 1 merge commit to the git history of main branch.

Elsewhere (@russellb ?) it was suggested that splitting in smaller pieces if they are logically separate (and they are here) is the right way of doing it. (And that's how I build my series in other projects over years.) I doubt you want to squash all of them into a one, but please confirm.

At the end of the day, it's not hard to squash the series and force-push. I just find it counterproductive to lose the context of the changes when dropping separate commit messages.

File: src/instructlab/utils.py
Comment (booxter) [reply src/instructlab/utils.py:301]: I will send a separate pylint cleanup PR later to handle these.

**Comment by leseb** [comment]
> > I don't think it's a good idea to add 35 commits + 1 merge commit to the git history of main branch.
> 
> Elsewhere (@russellb ?) it was suggested that splitting in smaller pieces if they are logically separate (and they are here) is the right way of doing it. (And that's how I build my series in other projects over years.) I doubt you want to squash all of them into a one, but please confirm.
> 
> At the end of the day, it's not hard to squash the series and force-push. I just find it counterproductive to lose the context of the changes when dropping separate commit messages.

💯

File: src/instructlab/utils.py
Comment (booxter) [reply src/instructlab/utils.py:387]: Strictly speaking, it is not related to mypy, but I will add this to the series to speed up review.

**Comment by booxter** [comment]
Changes since the previous version:

- dropped the patch that attempted to handle non-hashable object used as key to LRU cache (replaced with ignore comment);
- added a patch to switch from open(Path) to Path.open();
- removed upper-cap for referencing version in requirements.txt.

**Comment by tiran** [comment]
> At the end of the day, it's not hard to squash the series and force-push. I just find it counterproductive to lose the context of the changes when dropping separate commit messages.

Personally I prefer to keep separate commits in the PR but land each PR as one commit in the git commit history. This approach ensures a clean git history, but keeps the history of a change in the PR. With merge commits, every commit works, has passed the tests, and allows to perform git bisect + test.

Our approach with several commits + merge commit for each PR can create a broken git history, because any of the 3, 5, or 30 commits in a PR can cause a breakage. For example commit 2 breaks a test, commit 4 fixes it, and commits 1-5 land in the git history. This makes `git bisect` for bug hunting impossible, because commit 2 and 3 introduce a false positive error. The article https://danielrotter.at/2023/10/28/git-bisect-and-the-importance-of-a-clean-history.html explains the concept of git bisect and clean history in more words.

InstructLab decided to go for merge commits with several sub commits. I have to accept it, but I don't have to like it. I pity the maintainer that has to bisect a complex bug in the future...

**Comment by booxter** [comment]
I can only agree that, ideally, each commit would be CI'd. But that's something that the review / CI platform the project chose supports. (E.g. gerrit doesn't have this problem since each patch is validated.)

As I said, I can oblige to whatever the decision of the core team is. I hold off because I get inconsistent signals.

---

BTW since we don't gate pre-merge, the repo will have breaking patches merged regardless (though fewer). One would have to build a serial validation/merge pipeline to avoid this.

**Review by leseb — DISMISSED** [review]

**Review by tiran — COMMENTED** [review]
The changes are fine.

I still think it is too noisy to land these changes as a 38 separate commits. It's way too noisy in the git history. IMO we don't need fine-grained history for code refactoring. After all these changes should not change behavior.

**Review by leseb — APPROVED** [review]
@booxter has been very patient with this large change, and the constant rebasing must have been painful. The PR has no conflicts, so I think we should free him now :slightly_smiling_face: It’s been a while already. Honestly, when I look at all the questionable commits that get merged these days, I cringe every time I check the commit history. @booxter did a great job with well-formatted commit messages. I don’t think we should be blocking him on the number of commits. Thanks!

**Comment by tiran** [comment]
I like to add another data point. I haven't mentioned this before, but I did not want to insinuate foul play. 

As an open source project, we welcome contributions from everybody. In return, we recognize and celebrate the work from our contributors. For some contributors, metrics such as contribution and activity statistics are important. I fear that merging 38 separate commits will skew the statistics and set bad precedence. To put the number in relation: Russel is currently the top contributor with 89 commits in total (main branch, excluding merge commits). Places 9 has 38 commits and place 10 has 33 commits.

To be clear, I don't think that the 38 commits in this PR were created out of malicious intent. You are doing a great job. I really want to have your fixes in, so we can improve our typing and code quality even further.

I hope my explanation shows that there are not only technical issues (bisect, bloated history, hard to revert), but also a human issue.

With this new information, would you be willing to compromise and squash your PR into a smaller number of commits? I'm no longer insisting on one commit, but would prefer something along the line of half a dozen to a dozen commits.

Given your technical skills and experience with typing, I would also like to nominate you to become a triager and eventually a maintainer -- if you are interested. I'll have to talk to some people and figure out how to initiate the process.

**Comment by booxter** [comment]
@tiran don't get me wrong, I am not frustrated about any particular rule or policy (though of course I have my opinions), I just want to see some clarity as to what is actually considered "the right way" in this project. Signals so far are diametrically different. :) This is a bit frustrated indeed. (Honestly, I didn't think this PR will get so contentious.)

I understand people watch stats. Ideally, people don't use these numbers to celebrate or judge others' impact. (Which is e.g. minuscule here.) But if this keeps people happy, so be it. I merged most of mypy patches into one; left unrelated changes outside; only one mypy specific patch is left separated because it's not run-of-the-mill one.

Let me know if this is enough.

**Review by bjhargrave — APPROVED** [review]

**Review by tiran — APPROVED** [review]

File: src/instructlab/utils.py
Comment (nathan-weinberg) [reply src/instructlab/utils.py:375]: Looking at this now and boy do I concur

**Pull request merged**
